### DISCUSSION TITLE
How to get notification from a udev rule on Hyprland?

### DISCUSSION THREAD

**Discussion by Xenorf** [dunst-project/dunst#1266] [Q&A] [answered]
Hello everyone,

I integrated dunst to my Hyprland setup and it works very well. The thing is I don't really understand how to notify a user running Hyprland (me) from another user like root or in a service or udev rule.

I tried to run the command with the following environment variables:

- DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
- WAYLAND_DISPLAY=wayland-1
- DISPLAY=:0

It just changed the error message that I am getting: Error receiving data: Connection reset by peer

Do you have any idea what I am doing wrong and if it is even possible to get this working on Hyprland?
Thank you!

**Comment by bynect** [comment, marked as answer]
Rather than dunst directly something like this should probably be done by the notify client, for example `notify-send`.

With a quick search I found this:
https://superuser.com/questions/342626/notify-send-to-other-user-on-the-same-system

But basically it tells you to change the user with sudo before sending the notification

**Reply by Xenorf** [reply]
Hey thank you for your answer.

Indeed, I managed to get it working by using the su command on top of exporting the variables but I didn't think it was the cleanest way to go.

**Reply by bynect** [reply]
glad you resolved.
do you send the notification when a new device is attached/removed? I also want to do this one day or another :)

**Reply by Xenorf** [reply]
Yes I wanted to get a notification when a usb key is attached and removed. The udev rule I am using looks like that:
`ACTION=="add", SUBSYSTEMS=="usb", ENV{DEVTYPE}=="disk", RUN+="/bin/sh -c 'su <username> -c /home/<username>/.local/bin/usbmonitor.sh\ plug'"`

The script I use exports the **DBUS_SESSION_BUS_ADDRESS** before running the dunstify command:
`DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus dunstify -I $icon "USB key plugged in"`

**Reply by bynect** [reply]
Hello, I was doing something similar to you today and this is the script I have come up for now. This imports automatically the env vars from the dunst process.

```bash
#!/bin/bash

# Get envvar from another process
import_env() {
	for var in "$@"; do
		IFS='=' read key val < <(egrep -z "$var" /proc/$pid/environ)
		printf -v "$key" %s "$val"
		[[ ${!key} ]] && export "$key"
	done
}

notify() {
	local title=$1 body=$2

	# Assume dunst is the notification daemon for everyone
	declare -a pids=$(pgrep dunst)

	for pid in "${pids[@]}"; do
		(( pid )) || continue
		import_env XAUTHORITY USER DISPLAY DBUS_SESSION_BUS_ADDRESS
		su $USER -p -c "notify-send '$title' '$body'"
	done
}

notify "Attached new device" "Name: $1\nDevpath: $2"
```

the rule at the moment is
```
SUBSYSTEM=="usb", ACTION=="add", RUN+="/home/nect/script/udev-notify.sh '$name' '%p'"
```

However I still haven't figure out how to get a human readable name. Do you perchance know how?
